fl4health/clients/basic_client.py

```diff
@@ -77,6 +77,10 @@ def __init__(
 
         self.initialized = False  # Whether or not the client has been setup
 
+        # Variable to store additional information to be added to initial log
+        # string. This variable is reset every time _handle_logging is called
+        self.add_to_initial_log_str = ""
```

Comment:
I like the idea of providing a way for child clients to inject logging of specific information a lot. I'm not opposed to the route you have taken here, but sometime managing, appending to, and clearing "state" like this string can get a little convoluted. I'll suggest another potential way to do this below and you can see if you like it
```
def get_client_specific_logs(self) -> str:
    """
    This function can be used to provide any client specific or state 
    For example, perhaps a client uses a special LR scheduler and wants to 
    each logging round.
    NOTE: Needs to be overridden by an inheriting client to produce a 
    Returns:
        str: The logging value to be added to the logging string
    """
    return ""
```
Then you would simply do
```
initial_log_str += self.get_client_specific_logs()  # Add user defined info
```
in `_handle_logging` and then you'd just have 
```
def update_before_epoch(self, epoch: int) -> None:
        self.nnunet_trainer.lr_scheduler.step(epoch)

def get_client_specific_logs(self) -> str:
    lr = self.optimizers["global"].param_groups[0]["lr"]
    return f" Current LR: {lr}"
```
in your nnUnetClient code. This feels slightly cleaner to me, but would like to hear your thoughts.

Reply:
I agree that this is cleaner, and then makes it more obvious to the user that this is an available functionality. I do have some thoughts though that I'd like to know your opinion on. 

First is that it might be unclear exactly when handle_logging is called and by extension when get_client_specific_logs is called. For example the user might not be clear on whether LR is the current lr that is about to be used, or the most recent LR that was used. An easy fix would be to commit to always calling handle logs as the last step. The one exception might be the update_after_epoch hook, but I've actually removed this and replaced it with update before epoch. Either way if we can commit to when _handle_logging will be invoked, then we can communicate that to the user and i think that addresses that issue.

My second thought is that if we are going to add a ```get_client_specific_logs method```, then we should also add a ```get_client_specific_reports``` method to stay consistent.

My third thought is that ```get_client_specific_logs``` doesn't really communicate to the user that what is return is just appended to the initial log string. Again we could just communicate this in the docstring, or perhaps change the method name, although it does seem a bit weird to allow a user to add to the initial log string, but not to the rest of the logs itself. One idea that comes to mind is to have ```get_client_specific_logs``` return two values, a string and maybe an optional list of logs to add after logging the losses and metrics.

```python
class LogLevel(Enum):
    ERROR = logging.ERROR
    INFO = logging.INFO
    ...

def get_client_specific_logs(self) -> Tuple[Optional[str], Optional[List[Tuple[LogLevel, str]]]]:
    logs = []
    lr = self.optimizers["global"].param_groups[0]["lr"]
    logs.append((INFO, 'some more info that i want on a seperate line)
    return f" Current LR: {lr}", logs
```

Overall i think there are more upsides and less downsides with your approach. Curious to know what you think about some of these ideas